# Business Analytics - Assignment 2  

**Assignment Points**: 100  
**Submission**: Provide your answers in this notebook and submit it via iLearn

- Where a question requires a written (text) solution provide your answer in Markdown in appropriate cells under each question.
- Comment out your print statements unless you are explicitly asked to use the print() function. 
- 5 marks will be deducted for printed outputs that are not asked for.

### About the Assignment

- Assignment 2 extends Assignment 1 on credit card applications. 


- For this assignment there are two files in the `data` folder `credit_record.csv` and `application_record.csv` where bank clients are related by the `ID` column.
- In `credit_record.csv` we have the following variables

| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| AMT_INCOME_TOTAL   | Annual income  |  |
| NAME_INCOME_TYPE   | Income Source |  |
| NAME_EDUCATION_TYPE   | Level of Education  |  |
| CODE_GENDER   | Applicant's Gender   |  |
| FLAG_OWN_CAR | Car Ownership |  | 
| CNT_CHILDREN | Number of Children | |
| FLAG_OWN_REALTY | Real Estate Ownership | | 
| NAME_FAMILY_STATUS | Relationship Status | | 
| NAME_HOUSING_TYPE | Housing Type | | 
| DAYS_BIRTH | No. of Days | Count backwards from current day (0), -1 means yesterday
| DAYS_EMPLOYED | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| FLAG_MOBIL | Mobile Phone Ownership | | 
| FLAG_WORK_PHONE | Work Phone Ownership | | 
| FLAG_PHONE | Landline Phone Ownership | | 
| FLAG_EMAIL | Landline Phone Ownership | | 
| OCCUPATION_TYPE | Occupation | | 
| CNT_FAM_MEMBERS | Count of Family Members | |



- In `credit_record.csv` we have the following variables


| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number | |
| MONTHS_BALANCE | Number of months in the past from now when STATUS is measured | 0 = current month, -1 = last month, -2 = two months ago, etc.|
| STATUS | Number of days a payment is past due | 0: 1-29 days past due 1: 30-59 days past due 2: 60-89 days overdue 3: 90-119 days overdue 4: 120-149 days overdue 5: Overdue or bad debts, write-offs for more than 150 days C: paid off that month X: No loan for the month |

---
---

### Task 1: Reading, Summarising and Cleaning Data (Total Marks: 30)



**Question 1.** 

1. Import the `application_record.csv` and `credit_record.csv` files from `data` folder into pandas DataFrames named `df_application` and `df_credit`, respectively. (1 mark)

2. How many rows are there in `df_application` and `df_credit`, respectively? Answer using both print() function and in Markdown text. (1 mark)

3. How many unique bank clients are there in `df_application` and `df_credit`? Answer using both print() function and in Markdown text. (1 mark)

4. Add the records from `df_credit` to `df_application` by merging the data from the two DataFrames on the `ID` column, and output the joint data into a new DataFrame named `df`. Hint: Use `merge` function from pandas by setting `how` parameter to `inner` (4 marks) 

5. How many rows and how many unique clients are there in `df`? (1 mark)

6. How are multiple rows for each `ID` in `df` different? Answer in Markdown text. (2 mark) 

(10 marks)


In [1]:
# ---- Question 1 part 1 ------
import pandas as pd
df_application = pd.read_csv('data/application_record.csv')
df_credit = pd.read_csv('data/credit_record.csv')

# ---- Question 1 part 2 ----
print('Amount of rows in application record:', len(df_application))
print('Amount of rows in credit record:', len(df_credit))

# ---- Question 1 part 3 ----
print('There are',df_application['ID'].nunique(), 'unique bank clients in the application record')
print('There are',df_credit['ID'].nunique(), 'unique bank clients in the credit record')

# ---- Question 1 part 4 ----
df = pd.merge(df_application, df_credit, on = 'ID', how = 'inner')

# ---- Question 1 part 5 ----
print('The amount of rows in df are', len(df),'and there are',df['ID'].nunique(), 'unique bank clients')
#print('There are',df['ID'].nunique(), 'unique bank clients in the df')

Amount of rows in application record: 438557
Amount of rows in credit record: 1048575
There are 438510 unique bank clients in the application record
There are 45985 unique bank clients in the credit record
The amount of rows in df are 777715 and there are 36457 unique bank clients


### Written answers for question 1

Task 2 Response: There are 438,557 rows in df_application and 1,048,575 rows in df_credit 

Task 3 Response :There are 438,510 unique bank clients in df_application and 45,985 unique bank clients in df_credit

##### Question 1 Part 6

Multiple rows of `ID` are different as they we have now joined the credit record data onto the application records through the IDs, linking the rows that share the same ID's in both the credit and application record. So some will have more information attached because they existed in both records which are now displayed in the current dataframe `df`. 

---

**Question 2.**

1. Change the values of `STATUS` in `df` according to the following mapping: {C, X, 0} -> 0 and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers. (2 marks)
2. Create a new numpy array named `list_of_defaults` containing *unique* `ID` numbers for the clients who have `STATUS` = 1 in any of the last 12 months in the dataset. (2 marks) 
3. Create a new DataFrame called `df_final` that contains the rows of `df` for which the `ID` are in `list_of_defaults`, keeping only one row for each `ID` (i.e. eliminate rows with duplicate `ID`s while keeping the first duplicate row). How many rows do you have in `df_final`? Answer using both print() function and in Markdown text. (Hint: find out about `isin()` function in pandas.) (2 marks)
4. Add a new column `y = 1` for all the rows in `df_final`. (1 marks)
5. Increase `df_final` to a total of 4,000 rows by adding rows from `df` with unique `ID`s (nonduplicated `ID`s) which are not in `list_of_defaults`. To do this start adding the rows from the beginning of `df`. (Hint: learn what `~`, i.e. tilde sign, does in pandas). (2 marks) 
6. Fill the missing values of `y` in `df_final` with zeros. Remove `STATUS` and `MONTHS_BALANCE` from `df_final`. How many clients with  overdue payments of more than 29 days and how many clients with less than 29 days overdue payments are there in `df_final`? Answer using both print() function and in Markdown text.(1 mark)

(10 marks)

In [2]:
# ---- Question 2 part 1 ----
import numpy as np
status_mapping = {'C':0,'X':0,'0':0, '1':1,'2':1,'3':1,'4':1,'5':1}
df['STATUS'] = df['STATUS'].map(status_mapping).astype(int)

#print('unique rows in df', len(df['ID'].unique()))

# ---- Question 2 part 2 ----
list_of_defaults = np.array(df['ID'].loc[(df['STATUS']==1)&(df['MONTHS_BALANCE']>=-12)].unique())
#print('defaults', len(list_of_defaults))

# ---- Question 2 part 3 ----
df_final = pd.DataFrame(df.loc[(df['ID'].isin(list_of_defaults))])
 #print('before dupes', len(df_final))
df_final.drop_duplicates(subset='ID', inplace=True, keep = 'first')
print('The amount of rows in df_final are', len(df_final))

# ---- Question 2 part 4 ----
df_final = df_final.assign(y=lambda x: 1)
df_final

# ---- Question 2 part 5 ----
add_rows = df[(~df['ID'].isin(df_final))&(~df['ID'].isin(list_of_defaults))].drop_duplicates(subset='ID',keep='first')
df_final = pd.concat([df_final, add_rows], ignore_index=True)
df_final= df_final.loc[:3999]

# ---- Question 2 part 6 ----
df_final['y'] = df_final['y'].fillna(0)
df_final['y'] = df_final['y'].astype(int)

df_final = df_final.drop(['STATUS'], axis=1 )
df_final = df_final.drop(['MONTHS_BALANCE'], axis=1)

# How many clients with overdue payments of more than 29 days and how many clients with 
# less than 29 days overdue payments are there in `df_final`? Answer using both print() function and in Markdown text.
print('Number of clients with overdue payments more than 29 days is: ', sum(df_final['y']==0))
print('Number of clients with overdue payments less than 29 days is: ', sum(df_final['y']==1))

The amount of rows in df_final are 1833
Number of clients with overdue payments more than 29 days is:  2167
Number of clients with overdue payments less than 29 days is:  1833


#### Question 2 
There are 1833 rows in df_final after removing duplicate rows.

The number of clients with overdue payments of more than 29 days is 2,167 whilst the number of clients with overdue payments under 29 days is 1,833.

<hr style="width:25%;margin-left:0;"> 

**Question 3**. 
- Delete `ID` column from `df_final` (1 marks)
- Of the remaining variables in `df_final` and assuming that `NAME_EDUCATION_TYPE` is the only ordinal variable, how many variable are of numeric and nominal types? Provide lists of all numeric and nominal variables. (6)
- Using an appropriate function find and comment on the missing values in `df_final`, i.e. how many variables and how many observations? (3 marks)   
(10 marks)

In [3]:
df_final = df_final.drop(['ID'], axis=1)


missing_values_col = df_final.isna().sum()
#print(missing_values_col)
#df_final

Numerical Variables:

- CNT_CHILDREN
- AMT_INCOME_TOTAL
- DAYS_BIRTH
- DAYS_EMPLOYED
- COUNT_FAM_MEMBERS

Nominal Vairbales:

- CODE_GENDER
- FLAG_OWN_CAR
- FLAG_OWN_REALTY
- NAME_INCOME_TYPE
- NAME_FAMILY_STATUS
- NAME_HOUSING_TYPE
- FLAG_MOBIL
- FLAG_WORK_PHONE
- FLAG_PHONE
- FLAG_EMAIL
- OCCUPATION_TYPE
- y

It seems that there are 3 columns containing null values, these columns are: `CNT_CHILDREN`, `NAME_EDUCATION_TYPE`, `OCCUPATION_TYPE`. `CNT_CHILDREN` has 74 null obeservations, whilst `NAME_EDUCATION_TYPE`,`OCCUPATION_TYPE` have both 1831 and 1167 null observations respectively. 

---
---

### Task 2: Imputing missing values and dealing with categorical features (Total Marks: 30)



**Question 4.** 
- Use an appropriate `pandas` function to impute missing values in `df_final` (10 marks)
    - Be careful when deciding which method to use to replace missing observations 
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes
- Briefly explain what you have done and why. (5 marks)

(Total: 15 marks)

In [4]:
df_final['CNT_CHILDREN'].fillna(df_final['CNT_CHILDREN'].mode()[0], inplace=True)
df_final['NAME_EDUCATION_TYPE'].fillna(df_final['NAME_EDUCATION_TYPE'].mode()[0], inplace=True)
df_final['OCCUPATION_TYPE'].fillna(df_final['OCCUPATION_TYPE'].mode()[0], inplace=True)
#df_final

What I have done is replace all the NaN values in each column with the mode of that column. Using the mode in the case is more applicable as using the mean presents complications. As mean is a numerical output (the average of all outputs) it would result in a decimal (float) for the `CNT_CHILDREN` column and would not work in the string columns of `NAME_EDUCATION_TYPE` and `OCCUPATION_TYPE`. Using the mode which is the most common response in the column makes it versatile for both numerical and nominal results, so it is more applicable in this case.


<hr style="width:25%;margin-left:0;"> 

**Question 5**. Convert the values in `NAME_EDUCATION_TYPE` as follows
- Lower secondary -> 1
- Secondary / secondary special -> 2
- Incomplete higher -> 3
- Higher education -> 4


(Total: 5 marks)  

In [5]:
education_mapping = {'Lower secondary':1,'Secondary / secondary special':2,'Incomplete higher':3, 'Higher education':4}
df_final['NAME_EDUCATION_TYPE'] = df_final['NAME_EDUCATION_TYPE'].map(education_mapping).astype(int)
#df_final

<hr style="width:25%;margin-left:0;"> 

**Question 6**. 

Add dummy variables to `df_final` for all of the nominal features which are currently stored as string (text). 
- Make sure to delete the original variables from the dataframe
- Drop the first column from each set of created dummy variable, i.e. for each feature



(Total: 10 marks)  

In [6]:
dummy_gender = pd.get_dummies(df_final['CODE_GENDER'], prefix='GENDER', prefix_sep='_',drop_first=True)
dummy_car = pd.get_dummies(df_final['FLAG_OWN_CAR'], prefix='OWN_CAR', prefix_sep='_',drop_first=True)
dummy_realty = pd.get_dummies(df_final['FLAG_OWN_REALTY'], prefix='OWN_REALTY', prefix_sep='_',drop_first=True)
dummy_income = pd.get_dummies(df_final['NAME_INCOME_TYPE'], drop_first=True)
dummy_family = pd.get_dummies(df_final['NAME_FAMILY_STATUS'], drop_first=True)
dummy_housing = pd.get_dummies(df_final['NAME_HOUSING_TYPE'], drop_first=True)
dummy_occupation = pd.get_dummies(df_final['OCCUPATION_TYPE'], drop_first=True)

df_final = df_final.drop(['CODE_GENDER','FLAG_OWN_CAR','FLAG_OWN_REALTY','NAME_INCOME_TYPE','NAME_FAMILY_STATUS','NAME_HOUSING_TYPE','OCCUPATION_TYPE'],axis=1)

df_final = df_final.join(dummy_gender)
df_final = df_final.join(dummy_car)
df_final = df_final.join(dummy_realty)
df_final = df_final.join(dummy_income)
df_final = df_final.join(dummy_family)
df_final = df_final.join(dummy_housing)
df_final = df_final.join(dummy_occupation)

#df_final.info()

---
---

### Task 3 Preparing X and y arrays (Total Marks: 10)

**Question 7**. 

- Create a numpy array named `y` from the `y` column of `df_final` making sure that the values of the array `y` are stored as integers (3 marks)   
- Create a numpy array named `X`  from all the remaining features in `df_final` (2 marks)   

(Total: 5 Marks)

In [7]:
y = np.ravel(df_final['y'], order = 'C')
X = df_final.drop(['y'], axis=1).to_numpy()

<hr style="width:25%;margin-left:0;"> 

**Question 8**. 

- Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% train and 30% test datasets (2.5 marks) 
    - Set random_state to 7 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 7, stratify = y)

sc = StandardScaler()
sc.fit(X_train)

X_train_scaled = sc.transform(X_train)
X_test_scaled = sc.transform(X_test)

---
---

### Task 4. Support Vector Classifier and Accuracies (Total Marks: 30)


**Question 9**. 

- Train a Support Vector Classifier on standardised data (3 marks)
    - Use `linear` kernel and set `random_state` to 7 (don't change any other parameters)
    - Compute and print training and test dataset accuracies
- Train another Support Vector Classifier on standardised data (3 marks)
    - Use `rbf` kernel and set `random_state` to 7 (don't change any other parameters)
    - Compute and print training and test dataset accuracies
- What can you say about the presence of nonlinearities in the dataset? (4 marks)

(Total: 10 marks)  

In [9]:
from sklearn.svm import SVC

# kernel set to linear
svm = SVC(kernel='linear', random_state=7)
svm.fit(X_train_scaled, y_train)
print(f'Linear kernel accuracy on train data = {svm.score(X_train_scaled, y_train):.3f}')
print(f'Linear kernel accuracy on test data = {svm.score(X_test_scaled, y_test):.3f}')

# kernel set to rbf
svm = SVC(kernel='rbf', random_state=7)
svm.fit(X_train_scaled, y_train)
print(f'RBF kernel accuracy on train data = {svm.score(X_train_scaled, y_train):.3f}')
print(f'RBF kernel accuracy on test data = {svm.score(X_test_scaled, y_test):.3f}')

Linear kernel accuracy on train data = 0.645
Linear kernel accuracy on test data = 0.634
RBF kernel accuracy on train data = 0.777
RBF kernel accuracy on test data = 0.723


Based on the returned accuracies after running a linear kernel and a radial basis function (rbf) kernel, there is evidence of nonlinearities in the dataset as the `rbf` kernel produced a higher accuracy. This means that the dataset is not linearly separable and is likely more closely clustered together as there was a misclassification of 27.7% in the `rbf` model in contrast to the 36.6% misclassification`linear` model. This would suggest that the whether a client will default is determined by a combination of the predictors in the X array and will be in different regions depending on on the client instead of having a linear result.

<hr style="width:25%;margin-left:0;"> 

**Question 10**

- Extract 2 linear principal components from the standardised features using an appropriate `sklearn` library (5 marks)
- Train a Support Vector Classifier on the computed principal components (5 marks) 
    - Use `rbf` kernel and set `random_state` to 7 (don't change any other parameters)
- Compute and print training and test dataset accuracies (5 marks)
- What can you say about the ability of the 2 principal components to compress the information contained in the features matrix `X`, and why? (5 marks)     


(Total: 20 marks)  

In [10]:
from sklearn.decomposition import PCA
pca = PCA(n_components = 2)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

svm = SVC(kernel='rbf', random_state=7)
svm.fit(X_train_pca, y_train)
print(f'PCA RBF kernel accuracy on train data = {svm.score(X_train_pca, y_train):.3f}')
print(f'PCA RBF kernel accuracy on test data = {svm.score(X_test_pca, y_test):.3f}')

PCA RBF kernel accuracy on train data = 0.582
PCA RBF kernel accuracy on test data = 0.577


---- provide your text answer here ----
